guides/customizing_saving_and_serialization.py

```diff
@@ -0,0 +1,328 @@
+np.testing.assert_equal(restored_model.built, True)
+
+"""
+### `get_compile_config()` and `compile_from_config()`
+
+These methods work together to save the information with which the model was compiled
+(optimizers, losses, etc.) and restore and re-compile the model with this information.
+
+Overriding these methods can be useful for compiling the restored model with custom
+optimizers, custom losses, etc., as these will need to be deserialized prior to calling
+`model.compile` in `compile_from_config()`.
+
+Let's take a look at an example of this.
+
+**Example:**
+"""
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+def small_square_sum_loss(y_true, y_pred):
+    loss = tf.math.squared_difference(y_pred, y_true)
+    loss = loss / 10.0
+    loss = tf.reduce_sum(loss, axis=1)
+    return loss
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+def mean_pred(y_true, y_pred):
+    return tf.reduce_mean(y_pred)
+
+
+@keras.saving.register_keras_serializable(package="my_custom_package")
+class ModelWithCustomCompile(keras.Model):
+    def __init__(self):
+        super().__init__()
+        self.dense1 = keras.layers.Dense(8, activation="relu")
+        self.dense2 = keras.layers.Dense(4, activation="softmax")
+
+    def call(self, inputs):
+        x = self.dense1(inputs)
+        return self.dense2(x)
+
+    def compile(self, optimizer, loss_fn, metrics):
+        super().compile(optimizer=optimizer, loss=loss_fn, metrics=metrics)
+        self.model_optimizer = optimizer
+        self.loss_fn = loss_fn
+        self.loss_metrics = metrics
+
+    def get_compile_config(self):
+        # These parameters will be serialized at saving time.
+        return {
+            "model_optimizer": self.model_optimizer,
+            "loss_fn": self.loss_fn,
+            "metric": self.loss_metrics,
+        }
+
+    def compile_from_config(self, config):
+        # Deserializes the compile parameters (important, since many are custom)
+        optimizer = keras.utils.deserialize_keras_object(config["model_optimizer"])
+        loss_fn = keras.utils.deserialize_keras_object(config["loss_fn"])
+        metrics = keras.utils.deserialize_keras_object(config["metric"])
+
+        # Calls compile with the deserialized parameters
+        self.compile(optimizer=optimizer, loss_fn=loss_fn, metrics=metrics)
+
+
+model = ModelWithCustomCompile()
+model.compile(
+    optimizer="SGD", loss_fn=small_square_sum_loss, metrics=["accuracy", mean_pred]
+)
+
+x = np.random.random((4, 8))
+y = np.random.random((4,))
+
+model.fit(x, y)
+
+model.save("custom_compile_model.keras")
+restored_model = keras.models.load_model("custom_compile_model.keras")
+
+np.testing.assert_equal(model.model_optimizer, restored_model.model_optimizer)
+np.testing.assert_equal(model.loss_fn, restored_model.loss_fn)
+np.testing.assert_equal(model.loss_metrics, restored_model.loss_metrics)
```

Comment:
Please include a quick conclusion:

https://github.com/keras-team/keras-io/blob/master/contributor_guide.md#introduction-and-conclusion